src/pdu.c

```diff
@@ -703,52 +737,125 @@ coap_pdu_parse_opt(coap_pdu_t *pdu) {
-      case COAP_OPTION_URI_PATH:      if (len > 255) goto bad;            break;
-      case COAP_OPTION_CONTENT_FORMAT:if (len > 2) goto bad;              break;
-      case COAP_OPTION_MAXAGE:        if (len > 4) goto bad;              break;
-      case COAP_OPTION_URI_QUERY:     if (len < 1 || len > 255) goto bad; break;
-      case COAP_OPTION_HOP_LIMIT:     if (len < 1 || len > 1) goto bad;   break;
-      case COAP_OPTION_ACCEPT:        if (len > 2) goto bad;              break;
-      case COAP_OPTION_LOCATION_QUERY:if (len > 255) goto bad;            break;
-      case COAP_OPTION_BLOCK2:        if (len > 3) goto bad;              break;
-      case COAP_OPTION_BLOCK1:        if (len > 3) goto bad;              break;
-      case COAP_OPTION_SIZE2:         if (len > 4) goto bad;              break;
-      case COAP_OPTION_PROXY_URI:    if (len < 1 || len > 1034) goto bad; break;
-      case COAP_OPTION_PROXY_SCHEME:  if (len < 1 || len > 255) goto bad; break;
-      case COAP_OPTION_SIZE1:         if (len > 4) goto bad;              break;
-      case COAP_OPTION_NORESPONSE:    if (len > 1) goto bad;              break;
-      default:
-        ;
+      else if (!coap_pdu_parse_opt_base(pdu, len)) {
+        goto bad;
       }
       continue;
 bad:
-      coap_log(LOG_DEBUG,
-               "coap_pdu_parse: %d.%02d: option %u has bad length %u\n",
-               pdu->code >> 5, pdu->code & 0x1F, pdu->max_opt, len);
+      coap_log(LOG_WARNING,
+            "coap_pdu_parse: %d.%02d: offset %u option %u has bad length %u\n",
+               pdu->code >> 5, pdu->code & 0x1F,
+               (int)(opt_last - pdu->token - pdu->token_length), pdu->max_opt,
+               len);
       good = 0;
     }
 
+    if (!good) {
+      /*
+       * Dump the options in the PDU for analysis, space separated except
+       * error options which are prefixed by *
+       * Two rows - hex and ascii (if printable)
+       */
+      static char outbuf[100];
+      uint32_t tlen;
+      int outbuflen;
+
+      opt = pdu->token + pdu->token_length;
+      length = pdu->used_size - pdu->token_length;
+      pdu->max_opt = 0;
+
+      snprintf(outbuf, sizeof(outbuf), "O: ");
+      while (length > 0 && *opt != COAP_PAYLOAD_START) {
+        coap_opt_t *opt_last = opt;
+        size_t optsize = next_option_safe(&opt, &length, &pdu->max_opt);
+        const uint32_t len =
+          optsize ? coap_opt_length((const uint8_t *)opt - optsize) : 0;
+        if (!optsize) {
+          opt = pdu->token + pdu->token_length +
+                pdu->used_size - pdu->token_length;
+          length = 0;
+          goto badp;
+        }
+        if (COAP_PDU_IS_SIGNALING(pdu)) {
+          if (!coap_pdu_parse_opt_csm(pdu, len))
+            goto badp;
+        }
+        else if (!coap_pdu_parse_opt_base(pdu, len)) {
+          goto badp;
+        }
+        tlen = opt - opt_last;
+        outbuflen = strlen(outbuf);
```

Comment:
I do not think that we have to iterate over outbuf for every option to get the current length from the result of `snprintf` (if between 0 and remaining - 1).

Reply:
The issue here is that snprintf() does not return the actual number of characters written into the buffer and so cannot easily be used to compute an offset.
````
       The functions snprintf() and vsnprintf() do not write  more  than  size
       bytes  (including  the trailing '\0').  If the output was truncated due
       to this limit then the return value is the number  of  characters  (not
       including the trailing '\0') which would have been written to the final
       string if enough space had been available.  Thus,  a  return  value  of
       size  or  more  means  that  the output was truncated.
````

Reply:
Of course. The cases result < 0 and result >= remaining_length would need to be handled explicitly.

Reply:
Do you want me to come up with a change - same thing in coap_show_pdu()?

Reply:
I will take another look tomorrow to see how complex this will get.
The good thing about the strlen-approach is that it is much more readable.

Reply:
I was thinking of going in the direction of a safe (perhaps inline) function.

Reply:
I think that the following (untested) pattern would be readable enough to replace the strlen/snprintf:

```c
static inline int
write_prefix(char **obp, size_t *len, const char *prf, size_t prflen) {
  if (*len < prflen) {
    return 0;
  }

  memcpy(*obp, prf, prflen);
  *obp += prflen;
  *len -= prflen;
  return 1;    
}

static inline int
write_char(char **obp, size_t *len, char c, int printable) {
  if (*len < 2) {
    return 0;
  }

  if (printable && isprint(c)) {
    const uint8_t hex[] = "0123456789ABCDEF";
    (*obp)[0] = hex[(c & 0xf0) >> 4];
    (*obp)[1] = hex[c & 0xf0];
  } else {
    (*obp)[0] = c;
    (*obp)[1] = ' ';
  }
  *obp += 2;
  *len -= 2;
  return 1;
}

int
coap_pdu_parse_opt(coap_pdu_t *pdu) {
  /* ... */


  if (!good) {
      static char outbuf[COAP_DEBUG_BUF_SIZE];
      uint32_t tlen;
      size_t outbuflen = sizeof(outbuf);
      size_t i;
      char *obp = outbuf;

      for (i = 0; i < 2; i++) {
        /* ... */

        while (write_prefix(&obp, &outbuflen, "O: ", 3) && length > 0 ...) {
          /* ... */
          ok = ok && write_prefix(&obp, &outbuflen, "*", 1);
            /* ... */
          }
          /* ... */
          while (tlen--) {
            if (i == 0) {
              ok = ok && write_char(&obp, &outbuflen, *opt_last++, 0);
            } else {
              ok = ok && write_char(&obp, &outbuflen, *opt_last++, 1);
            }
          }
          /* ... */
        }
     }
  }
}
```

Reply:
Thanks, will check it through and update code.

Reply:
Change implemented (couple of fixes) and pushed.

Reply:
Great, and I like the implemented way to make sure we have space for the terminating zero. Can we put in a comment where the '\000' is inserted that the last successful write_* function made sure that prflen+1 bytes were left?

Reply:
Code updated and pushed.

Reply:
Thanks, ready to merge!